details/input.md

```diff
@@ -0,0 +1,172 @@
+# Interoperable Private Attribution wire format
+
+
+This documents provides clarification on the format IPA parties use to submit queries. 
+
+## Query
+
+IPA query consists of a mix of source and trigger events obtained from one or many source and trigger websites. 
+**Query size** is determined by the number of events included within a single query request. 
+
+It is desirable for report collectors to submit large queries as it brings more utility and saves cost, 
+therefore it makes sense to optimize the query format on the wire.
+
+The following sections propose the format that is space-optimized at the expense of being more complicated to assemble. 
+
+## Assumptions
+
+* Report collector use HTTP over TLS to send queries to helper party networks.
+* Number of events within a single query is between $10^6$ and $10^9$. 
+* Number of unique source and trigger websites is significantly lower than total number of events in the input set.
+
+
+## Format considerations
+
+It is worth to look at a single event first. To be concrete, this assumes match key to be a 40-bit length byte string,
+but this does not change the fundamentals. It is described in more details 
+[here](https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#generating-source-and-trigger-reports-by-the-report-collector).
+
+A single event consists of encrypted replicated shares of match key share, replicated shares of event data and additional 
+authenticated data. Event data varies depending on whether it is a source or a trigger event. Authenticated data is used
+to decrypt the shares of match key.
+
+Two things that consume the most space on the wire are **site registrable domain** and **match key provider origin**. 
+Both are ASCII strings, potentially large, that each event must refer to in order for helper parties to correctly obtain 
+the plain text shares.
+
+The biggest savings from the custom format come from making each query to carry only one copy of unique site domain
+and match key provider origin strings. This proposal suggests building two lookup tables (one for each entity) on the 
+caller site
```

Comment:
Given that source queries have a single source site and trigger queries have a single trigger site, you could start by indicating the type of query (which can be implicit or part of the query creation step), then you can have two tables: one for the "same" side (source configurations for source queries, trigger configurations for trigger queries) and one for the "other" side (the converse).  Then you could concatenate the two tables, index rows starting from zero and refer to the configurations.

Each row in the table would then effectively be a configuration that lists:

1. Site: length prefixed ASCII; 1 byte length.  Optimization hack: length = 0 copies the previous value.
2. Epoch: 2 bytes.
3. Key identifier: 1 byte.

There are three implied values that fill out the common stuff:

1. (implied) Event type is inferred from the table type, so this is effectively run-length encoded.
2. (implied) The match key provider should be the same for all events, so that can be part of query configuration.  
3. (implied) The helper party should know its own name, so that can be omitted completely.

Indexing into this table shouldn't take too many bytes.  But I don't think that a 1 byte is going to work out in all cases.  But the table size is known before you start processing individual items, so we can make the index size based on the table size ($\lceil log_2(t)/8 \rceil$).